lib/web/fetch/index.js

```diff
@@ -120,8 +122,49 @@ class Fetch extends EE {
   }
 }
 
+// publishIfSubscribed will check if there is context if so,
+// it will add error in the case of an error before publishing
+// the event if not it will return right away
+function publishIfSubscribed (name, context, error) {
+  if (!context) return
+  if (error) context.error = error
+  channels[name].publish(context)
+}
+
+// handlePromise will handle both fulfilled/rejected promises
+// from p and will publish events accordingly
+function handlePromise (p, context) {
+  return p.promise.then(
+    (success) => {
+      channels.asyncStart.publish(context)
+      channels.asyncEnd.publish(context)
+      return success
+    },
+    (failure) => {
+      context.error = failure
+      channels.asyncStart.publish(context)
+      channels.asyncEnd.publish(context)
+      return Promise.reject(failure)
+    })
+}
+
+// checking if we have subscribers at the beginning of the fetch call
+// this will set context if there are subscribers to any of the channels
+function subscribersCheck (input, init) {
+  const channelNames = ['start', 'end', 'asyncStart', 'asyncEnd', 'error']
+  for (let i = 0; i < channelNames.length; i++) {
+    const channelsSubscribers = channels[channelNames[i]]
+    if (channelsSubscribers.hasSubscribers) {
+      return { input, init, result: null, error: null }
+    }
+  }
+  return false
+}
```

Comment:
this function is highly unoptimized! 

you instantiate the array on every call
you know that there are only 5 possible values, yet you still need to make .length call and use a for loop?
return value is polymorphica as it returns an object and in other cases false

Reply:
probably easier and faster just to instantiate the context instead of calling this function...

Reply:
Ok, instead we will check just to see if the `start` channel has subscribers, if so we will set context and publish all events based on if the context has been set. I do think we should still only set context if we have a subscriber to the `start` channel, otherwise we are allocating memory for all function calls even those without any subscribers.

Reply:
Actually the memory footprint should be low because the attributes are by reference.

Reply:
How about this:

```js
function subscribersCheck (input, init) {
  const hasSubscribers = channels.start.hasSubscribers ||
    channels.end.hasSubscribers ||
    channels.asyncStart.hasSubscribers ||
    channels.asyncEnd.hasSubscribers ||
    channels.error.hasSubscribers

  if (!hasSubscribers) return

  return {
    input,
    init,
    result: null,
    error: null
  }
}
```

Reply:
@Qard right now I have the below which just checks if start channel has been subscribed to, let me know and I can update to your suggestion.

```
function subscribersCheck (input, init) {
  if (channels.start.hasSubscribers) {
    return { input, init, result: null, error: null }
  }
  return false
}
```

Reply:
You will need to check if _any_ of the channels have subscribers as it's totally valid to no listen to the start channel but to listen to others.